{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a520ec9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import Entrez\n",
    "from Bio import Medline\n",
    "from Bio import SeqIO\n",
    "import random\n",
    "import requests\n",
    "import re\n",
    "from tqdm import tqdm\n",
    "import nltk\n",
    "import time\n",
    "from itertools import cycle, islice\n",
    "from xml.etree import ElementTree\n",
    "import xmltodict\n",
    "import openai"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "96bfec28",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_document(document_ID):\n",
    "    query = str(document_ID)\n",
    "    def search(TERM):\n",
    "        Entrez.email = 'random@example.com'\n",
    "        handle = Entrez.esearch(db='pubmed', sort='relevance', retmax=1, retmode='text', term=TERM)\n",
    "        result = Entrez.read(handle)\n",
    "        ids = result['IdList']\n",
    "        handle = Entrez.efetch(db='pubmed', sort='relevance', retmode='text', rettype='medline', id=ids)\n",
    "        records = Medline.parse(handle)\n",
    "        return records\n",
    "\n",
    "\n",
    "    for record in search(query):\n",
    "        title = record.get('TI','')\n",
    "        abstract = record.get('AB','')\n",
    "        ID = record.get('PMID','')\n",
    "        mesh = record.get('MH','')\n",
    "        other_terms = record.get('OT','')\n",
    "\n",
    "    document = title + ' ' + abstract\n",
    "    with open('document.txt','w') as f:\n",
    "        f.write(document)\n",
    "    return ID, title, abstract, document"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "50b67d0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pubtator(document_ID):\n",
    "    FORMAT = 'biocxml'\n",
    "    TYPE = \"pmids\"\n",
    "    BIOCONCEPTS = 'gene,disease'\n",
    "    url = f\"https://www.ncbi.nlm.nih.gov/research/pubtator-api/publications/export/{FORMAT}?{TYPE}={document_ID}&concepts={BIOCONCEPTS}\"\n",
    "    response = requests.get(url)\n",
    "    time.sleep(1.5)\n",
    "    doc  = ElementTree.fromstring(response.content)\n",
    "    tree = ElementTree.ElementTree(doc)\n",
    "    tree.write('content.xml',encoding = 'utf-8')\n",
    "    root = tree.getroot()\n",
    "    doc = root[3]\n",
    "    passage = doc[1:]\n",
    "    TEXT = []\n",
    "    ELEMENT = []\n",
    "    IDENTIFIER = []\n",
    "    TYPE = []\n",
    "    for i in passage:\n",
    "        for text in i.iterfind('text'):\n",
    "            text = text.text\n",
    "            TEXT.append(text)\n",
    "        for annotation in i.iterfind('annotation'):\n",
    "            for text in annotation.iterfind('text'):\n",
    "                element = text.text\n",
    "                ELEMENT.append(element)\n",
    "            infos = annotation.findall('infon')\n",
    "            try: \n",
    "                identifier = infos[0].text\n",
    "            except:\n",
    "                identifier = ''\n",
    "            IDENTIFIER.append(identifier.replace('MESH:',''))\n",
    "            try:\n",
    "                typex = infos[1].text\n",
    "            except:\n",
    "                typex = infos[0].text\n",
    "            TYPE.append(typex)\n",
    "    df = pd.DataFrame(data=zip(ELEMENT,IDENTIFIER,TYPE), columns=['element','identifier','type'])\n",
    "    df.identifier = df.identifier.replace('Disease',np.nan)\n",
    "    df = df.drop_duplicates('identifier')\n",
    "    disease_df = df[df.type == 'Disease']\n",
    "    gene_df = df[df.type == 'Gene']\n",
    "    return gene_df, disease_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c921fb80",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pubtator_offline(document_ID):\n",
    "    with open('../DATI ABSTRACT/filtered_data/anns.txt','r') as f:\n",
    "        data = f.read().split('\\n')\n",
    "        data = [i for i in data if i]\n",
    "        data = [i for i in data if document_ID in i]\n",
    "        ELEMENT = []\n",
    "        IDENTIFIER = []\n",
    "        TYPE = []\n",
    "        for i in range(len(data)):\n",
    "            x = data[i].split('\\t')\n",
    "            ELEMENT.append(x[3])\n",
    "            IDENTIFIER.append(x[5])\n",
    "            TYPE.append(x[4])\n",
    "    df = pd.DataFrame(data=zip(ELEMENT,IDENTIFIER,TYPE), columns=['element','identifier','type'])\n",
    "    df.identifier = df.identifier.replace('Disease',np.nan)\n",
    "    df = df.drop_duplicates('identifier')\n",
    "    disease_df = df[df.type == 'Disease']\n",
    "    gene_df = df[df.type == 'Gene']\n",
    "    return gene_df, disease_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1428d1df",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pairs(gene_df, disease_df):\n",
    "    a = list(gene_df[['element', 'identifier']].itertuples(index=False, name=None))\n",
    "    b = list(disease_df[['element','identifier']].itertuples(index=False, name=None))\n",
    "    pairs = list(set([(i, j) for i in a for j in b]))\n",
    "    return pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "8e3ca5ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_associations_openai(document, document_id, pairs):\n",
    "    auth_params = {\"email\": \"salvatoredanilopalumbo@gmail.com\",\"password\" : \"PVFXYXpkV1p1VkdkeUVETTFrak4wVm1ibGQyY3BSR1JCNVVTTTkwVE1sa1RCUkVacE4zWmw1V1owWlRPMUFUTXlRWFp1VjJaemxHWg==\"}\n",
    "    api_host = \"https://www.disgenet.org/api\"\n",
    "    VOCABULARY = 'mesh'\n",
    "    SOURCE = 'CURATED'\n",
    "    \n",
    "    openai.api_key = \"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\"\n",
    "    \n",
    "    # engine = \"text-davinci-001\"\n",
    "    engine = 'text-curie-001'\n",
    "\n",
    "    \n",
    "    temperature = 0\n",
    "    max_tokens = 500\n",
    "    top_p = 1\n",
    "    frequency_penalty = 0\n",
    "    presence_penalty = 0\n",
    "    \n",
    "    \n",
    "    DATA = pd.DataFrame(columns = ['geneId','diseaseId'])\n",
    "    gene_id = []\n",
    "    disease_id = []\n",
    "    disease_umls = []\n",
    "    \n",
    "    for j in pairs:\n",
    "        prompt = f\"According to this text:\\n\\n{document}\\n\\nIs {j[0][0]} associated with {j[1][0]}?\"\n",
    "        response = openai.Completion.create(engine=engine,prompt=prompt,temperature=temperature,max_tokens=max_tokens,top_p=top_p,frequency_penalty=frequency_penalty,\n",
    "                    presence_penalty=presence_penalty)\n",
    "        time.sleep(1.5)\n",
    "        print(prompt)\n",
    "        for i in response['choices']:\n",
    "            p = str(i['text'])\n",
    "            p = p.replace('\\n','')\n",
    "            print(p)\n",
    "            if p.startswith('Yes'):\n",
    "                gene_id.append(j[0][1])\n",
    "                disease_id.append(j[1][1])\n",
    "                # print(f'{j[0][0]} - {j[0][1]}')\n",
    "                # print(f'{j[1][0]} - {j[1][1]}')\n",
    "    DATA.geneId = gene_id\n",
    "    DATA.diseaseId = disease_id\n",
    "    return DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e5025fc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb557612",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17e3d94c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfbe3d8f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e2c238e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dee981f6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2e6ad7f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "18b46e5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "According to this text:\n",
      "\n",
      "Strong association of a novel Tau promoter haplotype in progressive supranuclear palsy. The microtubule associated protein, tau, is found in fibrillar lesions that characterise progressive supranuclear palsy (PSP) and related tauopathies. Mutations in the tau gene in frontotemporal dementia with parkinsonism linked to chromosome 17 (FTDP-17) and genetic association of the H1 haplotype of the tau gene with PSP has firmly established a direct role for tau in disease pathogenesis. However, the functional significance of the tau genetic association in PSP is unknown. We analysed the tau gene promoter sequence and identified two novel single nucleotide polymorphisms. Here we report the genetic association of a novel tau promoter haplotype with PSP which may influence tau transcription.\n",
      "\n",
      "Is microtubule associated protein associated with supranuclear palsy?\n",
      "Yes, the microtubule associated protein, tau, is found in fibrillar lesions that characterise progressive supranuclear palsy (PSP) and related tauopathies.\n",
      "According to this text:\n",
      "\n",
      "Strong association of a novel Tau promoter haplotype in progressive supranuclear palsy. The microtubule associated protein, tau, is found in fibrillar lesions that characterise progressive supranuclear palsy (PSP) and related tauopathies. Mutations in the tau gene in frontotemporal dementia with parkinsonism linked to chromosome 17 (FTDP-17) and genetic association of the H1 haplotype of the tau gene with PSP has firmly established a direct role for tau in disease pathogenesis. However, the functional significance of the tau genetic association in PSP is unknown. We analysed the tau gene promoter sequence and identified two novel single nucleotide polymorphisms. Here we report the genetic association of a novel tau promoter haplotype with PSP which may influence tau transcription.\n",
      "\n",
      "Is Tau associated with supranuclear palsy?\n",
      "Yes, the novel Tau promoter haplotype is associated with PSP.\n",
      "According to this text:\n",
      "\n",
      "Strong association of a novel Tau promoter haplotype in progressive supranuclear palsy. The microtubule associated protein, tau, is found in fibrillar lesions that characterise progressive supranuclear palsy (PSP) and related tauopathies. Mutations in the tau gene in frontotemporal dementia with parkinsonism linked to chromosome 17 (FTDP-17) and genetic association of the H1 haplotype of the tau gene with PSP has firmly established a direct role for tau in disease pathogenesis. However, the functional significance of the tau genetic association in PSP is unknown. We analysed the tau gene promoter sequence and identified two novel single nucleotide polymorphisms. Here we report the genetic association of a novel tau promoter haplotype with PSP which may influence tau transcription.\n",
      "\n",
      "Is microtubule associated protein associated with dementia?\n",
      "Yes, tau is associated with dementia.\n",
      "According to this text:\n",
      "\n",
      "Strong association of a novel Tau promoter haplotype in progressive supranuclear palsy. The microtubule associated protein, tau, is found in fibrillar lesions that characterise progressive supranuclear palsy (PSP) and related tauopathies. Mutations in the tau gene in frontotemporal dementia with parkinsonism linked to chromosome 17 (FTDP-17) and genetic association of the H1 haplotype of the tau gene with PSP has firmly established a direct role for tau in disease pathogenesis. However, the functional significance of the tau genetic association in PSP is unknown. We analysed the tau gene promoter sequence and identified two novel single nucleotide polymorphisms. Here we report the genetic association of a novel tau promoter haplotype with PSP which may influence tau transcription.\n",
      "\n",
      "Is Tau associated with dementia?\n",
      "Yes, Tau is associated with dementia.\n",
      "According to this text:\n",
      "\n",
      "Strong association of a novel Tau promoter haplotype in progressive supranuclear palsy. The microtubule associated protein, tau, is found in fibrillar lesions that characterise progressive supranuclear palsy (PSP) and related tauopathies. Mutations in the tau gene in frontotemporal dementia with parkinsonism linked to chromosome 17 (FTDP-17) and genetic association of the H1 haplotype of the tau gene with PSP has firmly established a direct role for tau in disease pathogenesis. However, the functional significance of the tau genetic association in PSP is unknown. We analysed the tau gene promoter sequence and identified two novel single nucleotide polymorphisms. Here we report the genetic association of a novel tau promoter haplotype with PSP which may influence tau transcription.\n",
      "\n",
      "Is microtubule associated protein associated with parkinsonism?\n",
      "Yes, mutations in the tau gene in frontotemporal dementia with parkinsonism linked to chromosome 17 (FTDP-17) and genetic association of the H1 haplotype of the tau gene with PSP have firmly established a direct role for tau in disease pathogenesis. However, the functional significance of the tau genetic association in PSP is unknown.\n",
      "According to this text:\n",
      "\n",
      "Strong association of a novel Tau promoter haplotype in progressive supranuclear palsy. The microtubule associated protein, tau, is found in fibrillar lesions that characterise progressive supranuclear palsy (PSP) and related tauopathies. Mutations in the tau gene in frontotemporal dementia with parkinsonism linked to chromosome 17 (FTDP-17) and genetic association of the H1 haplotype of the tau gene with PSP has firmly established a direct role for tau in disease pathogenesis. However, the functional significance of the tau genetic association in PSP is unknown. We analysed the tau gene promoter sequence and identified two novel single nucleotide polymorphisms. Here we report the genetic association of a novel tau promoter haplotype with PSP which may influence tau transcription.\n",
      "\n",
      "Is Tau associated with parkinsonism?\n",
      "Yes, Tau is associated with Parkinson's disease.\n",
      "According to this text:\n",
      "\n",
      "Strong association of a novel Tau promoter haplotype in progressive supranuclear palsy. The microtubule associated protein, tau, is found in fibrillar lesions that characterise progressive supranuclear palsy (PSP) and related tauopathies. Mutations in the tau gene in frontotemporal dementia with parkinsonism linked to chromosome 17 (FTDP-17) and genetic association of the H1 haplotype of the tau gene with PSP has firmly established a direct role for tau in disease pathogenesis. However, the functional significance of the tau genetic association in PSP is unknown. We analysed the tau gene promoter sequence and identified two novel single nucleotide polymorphisms. Here we report the genetic association of a novel tau promoter haplotype with PSP which may influence tau transcription.\n",
      "\n",
      "Is microtubule associated protein associated with PSP?\n",
      "Yes, the microtubule associated protein, tau, is found in fibrillar lesions that characterise progressive supranuclear palsy (PSP) and related tauopathies.\n",
      "According to this text:\n",
      "\n",
      "Strong association of a novel Tau promoter haplotype in progressive supranuclear palsy. The microtubule associated protein, tau, is found in fibrillar lesions that characterise progressive supranuclear palsy (PSP) and related tauopathies. Mutations in the tau gene in frontotemporal dementia with parkinsonism linked to chromosome 17 (FTDP-17) and genetic association of the H1 haplotype of the tau gene with PSP has firmly established a direct role for tau in disease pathogenesis. However, the functional significance of the tau genetic association in PSP is unknown. We analysed the tau gene promoter sequence and identified two novel single nucleotide polymorphisms. Here we report the genetic association of a novel tau promoter haplotype with PSP which may influence tau transcription.\n",
      "\n",
      "Is Tau associated with PSP?\n",
      "Yes, the novel Tau promoter haplotype is associated with PSP.\n",
      "According to this text:\n",
      "\n",
      "Strong association of a novel Tau promoter haplotype in progressive supranuclear palsy. The microtubule associated protein, tau, is found in fibrillar lesions that characterise progressive supranuclear palsy (PSP) and related tauopathies. Mutations in the tau gene in frontotemporal dementia with parkinsonism linked to chromosome 17 (FTDP-17) and genetic association of the H1 haplotype of the tau gene with PSP has firmly established a direct role for tau in disease pathogenesis. However, the functional significance of the tau genetic association in PSP is unknown. We analysed the tau gene promoter sequence and identified two novel single nucleotide polymorphisms. Here we report the genetic association of a novel tau promoter haplotype with PSP which may influence tau transcription.\n",
      "\n",
      "Is Tau associated with tauopathies?\n",
      "Yes, the novel tau promoter haplotype is associated with PSP.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "According to this text:\n",
      "\n",
      "Strong association of a novel Tau promoter haplotype in progressive supranuclear palsy. The microtubule associated protein, tau, is found in fibrillar lesions that characterise progressive supranuclear palsy (PSP) and related tauopathies. Mutations in the tau gene in frontotemporal dementia with parkinsonism linked to chromosome 17 (FTDP-17) and genetic association of the H1 haplotype of the tau gene with PSP has firmly established a direct role for tau in disease pathogenesis. However, the functional significance of the tau genetic association in PSP is unknown. We analysed the tau gene promoter sequence and identified two novel single nucleotide polymorphisms. Here we report the genetic association of a novel tau promoter haplotype with PSP which may influence tau transcription.\n",
      "\n",
      "Is microtubule associated protein associated with tauopathies?\n",
      "Yes, the microtubule associated protein, tau, is found in fibrillar lesions that characterise progressive supranuclear palsy (PSP) and related tauopathies.\n"
     ]
    }
   ],
   "source": [
    "x = '11578815'\n",
    "RESULT = pd.DataFrame(columns = ['geneId','diseaseId'])\n",
    "ID, title, abstract, document = parse_document(x)\n",
    "gene_df, disease_df = pubtator(ID)\n",
    "pairs = get_pairs(gene_df, disease_df)\n",
    "result = check_associations_openai(document, ID, pairs)\n",
    "RESULT = RESULT.append(result, ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "06b0c277",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>element</th>\n",
       "      <th>identifier</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Tau</td>\n",
       "      <td>4137</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>microtubule associated protein</td>\n",
       "      <td>51115</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          element identifier  type\n",
       "0                             Tau       4137  Gene\n",
       "2  microtubule associated protein      51115  Gene"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "0868520f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{lll}\n",
      "\\toprule\n",
      "                       element & identifier & type \\\\\n",
      "\\midrule\n",
      "                           Tau &       4137 & Gene \\\\\n",
      "microtubule associated protein &      51115 & Gene \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(gene_df.to_latex(index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "314b8637",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>element</th>\n",
       "      <th>identifier</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>supranuclear palsy</td>\n",
       "      <td>D013494</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PSP</td>\n",
       "      <td>D011030</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tauopathies</td>\n",
       "      <td>D024801</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>dementia</td>\n",
       "      <td>D003704</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>parkinsonism</td>\n",
       "      <td>D010302</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              element identifier     type\n",
       "1  supranuclear palsy    D013494  Disease\n",
       "5                 PSP    D011030  Disease\n",
       "6         tauopathies    D024801  Disease\n",
       "8            dementia    D003704  Disease\n",
       "9        parkinsonism    D010302  Disease"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "4aa0bca0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{lll}\n",
      "\\toprule\n",
      "           element & identifier &    type \\\\\n",
      "\\midrule\n",
      "supranuclear palsy &    D013494 & Disease \\\\\n",
      "               PSP &    D011030 & Disease \\\\\n",
      "       tauopathies &    D024801 & Disease \\\\\n",
      "          dementia &    D003704 & Disease \\\\\n",
      "      parkinsonism &    D010302 & Disease \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(disease_df.to_latex(index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "495c5dd3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(('microtubule associated protein', '51115'), ('supranuclear palsy', 'D013494'))\n",
      "(('Tau', '4137'), ('supranuclear palsy', 'D013494'))\n",
      "(('microtubule associated protein', '51115'), ('dementia', 'D003704'))\n",
      "(('Tau', '4137'), ('dementia', 'D003704'))\n",
      "(('microtubule associated protein', '51115'), ('parkinsonism', 'D010302'))\n",
      "(('Tau', '4137'), ('parkinsonism', 'D010302'))\n",
      "(('microtubule associated protein', '51115'), ('PSP', 'D011030'))\n",
      "(('Tau', '4137'), ('PSP', 'D011030'))\n",
      "(('Tau', '4137'), ('tauopathies', 'D024801'))\n",
      "(('microtubule associated protein', '51115'), ('tauopathies', 'D024801'))\n"
     ]
    }
   ],
   "source": [
    "for i in pairs:\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "248c7eaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneId</th>\n",
       "      <th>diseaseId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>51115</td>\n",
       "      <td>D013494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4137</td>\n",
       "      <td>D013494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>51115</td>\n",
       "      <td>D003704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4137</td>\n",
       "      <td>D003704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>51115</td>\n",
       "      <td>D010302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4137</td>\n",
       "      <td>D010302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>51115</td>\n",
       "      <td>D011030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4137</td>\n",
       "      <td>D011030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4137</td>\n",
       "      <td>D024801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>51115</td>\n",
       "      <td>D024801</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  geneId diseaseId\n",
       "0  51115   D013494\n",
       "1   4137   D013494\n",
       "2  51115   D003704\n",
       "3   4137   D003704\n",
       "4  51115   D010302\n",
       "5   4137   D010302\n",
       "6  51115   D011030\n",
       "7   4137   D011030\n",
       "8   4137   D024801\n",
       "9  51115   D024801"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RESULT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "045216b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{ll}\n",
      "\\toprule\n",
      "geneId & diseaseId \\\\\n",
      "\\midrule\n",
      " 51115 &   D013494 \\\\\n",
      "  4137 &   D013494 \\\\\n",
      " 51115 &   D003704 \\\\\n",
      "  4137 &   D003704 \\\\\n",
      " 51115 &   D010302 \\\\\n",
      "  4137 &   D010302 \\\\\n",
      " 51115 &   D011030 \\\\\n",
      "  4137 &   D011030 \\\\\n",
      "  4137 &   D024801 \\\\\n",
      " 51115 &   D024801 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(RESULT.to_latex(index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df29a7a6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dac8dac",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa89a24d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db89215a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3e9ee8f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "16f38778",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_document = 'documents_in_filtered_data'\n",
    "documents_input = pd.read_csv(f'../{input_document}.csv')\n",
    "LIST_ID = list(set([str(i) for i in documents_input.pmid.tolist()]))\n",
    "len(LIST_ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "896abe1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████| 100/100 [21:18<00:00, 12.79s/it]\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: '4681|100532736'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_14028/3960228304.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      8\u001b[0m     \u001b[0mresult\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcheck_associations_openai\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdocument\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mID\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mpairs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      9\u001b[0m     \u001b[0mRESULT\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mRESULT\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mignore_index\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 10\u001b[1;33m \u001b[0mRESULT\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgeneId\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mRESULT\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgeneId\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mint64\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\generic.py\u001b[0m in \u001b[0;36mastype\u001b[1;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[0;32m   5813\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   5814\u001b[0m             \u001b[1;31m# else, only a single dtype is given\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 5815\u001b[1;33m             \u001b[0mnew_data\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_mgr\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   5816\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_constructor\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mnew_data\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__finalize__\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"astype\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   5817\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\internals\\managers.py\u001b[0m in \u001b[0;36mastype\u001b[1;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[0;32m    416\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    417\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mastype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mT\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mbool\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mFalse\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mstr\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m\"raise\"\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m->\u001b[0m \u001b[0mT\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 418\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"astype\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    419\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    420\u001b[0m     def convert(\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\internals\\managers.py\u001b[0m in \u001b[0;36mapply\u001b[1;34m(self, f, align_keys, ignore_failures, **kwargs)\u001b[0m\n\u001b[0;32m    325\u001b[0m                     \u001b[0mapplied\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mb\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mf\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    326\u001b[0m                 \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 327\u001b[1;33m                     \u001b[0mapplied\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mb\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    328\u001b[0m             \u001b[1;32mexcept\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mTypeError\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mNotImplementedError\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    329\u001b[0m                 \u001b[1;32mif\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mignore_failures\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\internals\\blocks.py\u001b[0m in \u001b[0;36mastype\u001b[1;34m(self, dtype, copy, errors)\u001b[0m\n\u001b[0;32m    589\u001b[0m         \u001b[0mvalues\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    590\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 591\u001b[1;33m         \u001b[0mnew_values\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mastype_array_safe\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    592\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    593\u001b[0m         \u001b[0mnew_values\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmaybe_coerce_values\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mnew_values\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\dtypes\\cast.py\u001b[0m in \u001b[0;36mastype_array_safe\u001b[1;34m(values, dtype, copy, errors)\u001b[0m\n\u001b[0;32m   1307\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1308\u001b[0m     \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1309\u001b[1;33m         \u001b[0mnew_values\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mastype_array\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1310\u001b[0m     \u001b[1;32mexcept\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mValueError\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1311\u001b[0m         \u001b[1;31m# e.g. astype_nansafe can fail on object-dtype of strings\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\dtypes\\cast.py\u001b[0m in \u001b[0;36mastype_array\u001b[1;34m(values, dtype, copy)\u001b[0m\n\u001b[0;32m   1255\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1256\u001b[0m     \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1257\u001b[1;33m         \u001b[0mvalues\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mastype_nansafe\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1258\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1259\u001b[0m     \u001b[1;31m# in pandas we don't store numpy str dtypes, so convert to object\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\core\\dtypes\\cast.py\u001b[0m in \u001b[0;36mastype_nansafe\u001b[1;34m(arr, dtype, copy, skipna)\u001b[0m\n\u001b[0;32m   1172\u001b[0m         \u001b[1;31m# work around NumPy brokenness, #1987\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1173\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0missubdtype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtype\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0minteger\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1174\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mlib\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mastype_intsafe\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0marr\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1175\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1176\u001b[0m         \u001b[1;31m# if we have a datetime/timedelta array of objects\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\pandas\\_libs\\lib.pyx\u001b[0m in \u001b[0;36mpandas._libs.lib.astype_intsafe\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: invalid literal for int() with base 10: '4681|100532736'"
     ]
    }
   ],
   "source": [
    "COUNTER = 4\n",
    "RESULT = pd.DataFrame(columns = ['geneId','diseaseId','diseaseUMLS'])\n",
    "for x in tqdm(LIST_ID):\n",
    "    ID, title, abstract, document = parse_document(x)\n",
    "    # gene_df, disease_df = pubtator(ID)\n",
    "    gene_df, disease_df = pubtator_offline(ID)\n",
    "    pairs = get_pairs(gene_df, disease_df)\n",
    "    result = check_associations_openai(document, ID, pairs)\n",
    "    RESULT = RESULT.append(result, ignore_index = True)\n",
    "RESULT.geneId = RESULT.geneId.astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "05103c2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "RESULT.to_csv(f'openai_results_no_conversion_{COUNTER}_{input_document}.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2447d3f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "auth_params = {\"email\": \"salvatoredanilopalumbo@gmail.com\",\"password\" : \"PVFXYXpkV1p1VkdkeUVETTFrak4wVm1ibGQyY3BSR1JCNVVTTTkwVE1sa1RCUkVacE4zWmw1V1owWlRPMUFUTXlRWFp1VjJaemxHWg==\"}\n",
    "api_host = \"https://www.disgenet.org/api\"\n",
    "VOCABULARY = 'mesh'\n",
    "DISEASE_UMLS = []\n",
    "\n",
    "s = requests.Session()\n",
    "r = s.post(api_host+'/auth/', data=auth_params)\n",
    "json_response = r.json()\n",
    "api_key = json_response.get(\"token\")\n",
    "\n",
    "\n",
    "for i in tqdm(RESULT.diseaseId):\n",
    "    if api_key:\n",
    "        s.headers.update({\"Authorization\": \"Bearer %s\" % api_key})\n",
    "        try:\n",
    "            gda_response = s.get(f\"{api_host}/gda/disease/{VOCABULARY}/{i}\").json()\n",
    "            DISEASE_UMLS.append(list(set([i['diseaseid'] for i in gda_response])))\n",
    "        except:\n",
    "            DISEASE_UMLS.append(np.nan)\n",
    "if s:\n",
    "    s.close()\n",
    "        \n",
    "RESULT['diseaseUMLS'] = DISEASE_UMLS\n",
    "RESULT.dropna(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c639324f",
   "metadata": {},
   "outputs": [],
   "source": [
    "RESULT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35a21963",
   "metadata": {},
   "outputs": [],
   "source": [
    "RESULT.to_csv(f'openai_results_{COUNTER}_unexploded_{input_document}.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d46d9cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "RESULT = RESULT.explode('diseaseUMLS')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c3d9d0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "RESULT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c612e0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "RESULT.to_csv(f'openai_results_{COUNTER}_{input_document}.csv',index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
